Python/Product/Analysis/LanguageServer/EditorFiles.cs

```diff
@@ -0,0 +1,237 @@
+                var part = _server.ProjectFiles.GetPart(uri);
+                _server.TraceMessage($"Received changes for {uri}");
+
+                var docVersion = Math.Max(doc.GetDocumentVersion(part), 0);
+                var fromVersion = Math.Max(@params.textDocument.version - 1 ?? docVersion, 0);
+
+                if (fromVersion > docVersion && @params.contentChanges?.Any(c => c.range == null) != true) {
+                    // Expected from version hasn't been seen yet, and there are no resets in this
+                    // change, so enqueue it for later.
+                    _server.TraceMessage($"Deferring changes for {uri} until version {fromVersion} is seen");
+                    lock (_pendingChanges) {
+                        _pendingChanges.Add(@params);
+                    }
+                    return;
+                }
+
+                var toVersion = @params.textDocument.version ?? (fromVersion + changes.Length);
+
+                doc.UpdateDocument(part, new DocumentChangeSet(
+                    fromVersion,
+                    toVersion,
+                    changes.Select(c => new DocumentChange {
+                        ReplacedSpan = c.range.GetValueOrDefault(),
+                        WholeBuffer = !c.range.HasValue,
+                        InsertedText = c.text
+                    })
+                ));
+
+                DidChangeTextDocumentParams? next = null;
+                lock (_pendingChanges) {
+                    var notExpired = _pendingChanges
+                        .Where(p => p.textDocument.version.GetValueOrDefault() >= toVersion)
+                        .OrderBy(p => p.textDocument.version.GetValueOrDefault())
+                        .ToArray();
+
+                    _pendingChanges.Clear();
+                    if (notExpired.Any()) {
+                        next = notExpired.First();
+                        _pendingChanges.AddRange(notExpired.Skip(1));
+                    }
+                }
+                if (next.HasValue) {
+                    DidChangeTextDocument(next.Value, null);
+                }
+            } finally {
+                if (enqueueAction != null) {
+                    _server.TraceMessage($"Applied changes to {uri}");
+                    enqueueAction(doc);
+                }
+            }
+        }
+
+        public void UpdateParseDiagnostics(VersionCookie vc, Uri documentUri) {
+            List<PublishDiagnosticsEventArgs> diags = null;
+
+            lock (_lock) {
+                var last = _parseBufferDiagnostics;
+
+                foreach (var kv in vc.GetAllParts(documentUri)) {
+                    var part = _server.ProjectFiles.GetPart(kv.Key);
+                    if (!last.TryGetValue(part, out var lastVersion) || lastVersion.Version < kv.Value.Version || _ignoreDiagnosticsVersion) {
+                        last[part] = kv.Value;
+                            diags = diags ?? new List<PublishDiagnosticsEventArgs>();
```

Comment:
Is the extra indentation here actually in the code? Or just showing up on GH for some reason?